File: atcoder/abc265B_explore.hs
-- uninhm
-- https://atcoder.jp/contests/abc265/tasks/abc265_b
-- implementation

import qualified Data.Map as M

import Control.Monad (replicateM)
import Data.Bool (bool)

getInts :: IO [Int]
getInts = map read . words <$> getLine

solve :: Int -> [Int] -> M.Map Int Int -> Int -> Bool
solve t [] _ _ = t > 0
solve t (x:a) bonuses i
  | t <= 0 = False
  | otherwise = solve (t - x + bonus) a bonuses (i+1)
  where
    bonus = M.findWithDefault 0 i bonuses

main :: IO ()
main = do
  [n, m, t] <- getInts
  a <- getInts
  bonuses <- M.fromList . map (\[a,b] -> (a,b)) <$> replicateM m getInts
  putStrLn $ bool "No" "Yes" $ solve t a bonuses 1

File: atcoder/abc265B_explore.cpp
// vicfred
// https://atcoder.jp/contests/abc265/tasks/abc265_b
// implementation
#include <iostream>
#include <vector>

using namespace std;

int main() {
  long long N, M, T;
  cin >> N >> M >> T;
  vector<int> A(N + 1);
  for (int i = 1; i <= N - 1; ++i) {
    cin >> A[i];
  }
  for (int i = 0; i < M; ++i) {
    int X, Y;
    cin >> X >> Y;
    A[X] -= Y;
  }
  bool valid = true;
  for (int i = 1; i <= N - 1; ++i) {
    T -= A[i];
    if (T <= 0) {
      valid = false;
    }
  }
  if (valid) {
    cout << "Yes" << endl;
  } else {
    cout << "No" << endl;
  }
  return 0;
}
